src/controller/ICDRegistrationDelegate.h

```diff
@@ -0,0 +1,80 @@
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *  @file
+ *    This file contains definitions for DeviceProxy for a device that's undergoing
+ *    commissioning process. The objects of this will be used by Controller applications
+ *    to interact with the device. The class provides mechanism to construct, send and receive
+ *    messages to and from the corresponding CHIP devices.
+ */
+
+#pragma once
+
+#include <cinttypes>
+
+#include <lib/support/Span.h>
+
+namespace chip {
+namespace Controller {
+
+/**
+ * @brief
+ * ICDRegistrationDelegate is an interface that provides a callback method for generating symmetric keys and reporting ICD
+ * registration readiness.
+ *
+ * This delegate is used by the commissioner to generate symmetric keys and complete ICD registration with the ICD
+ * device and to report when the device is ready for subscription.
+ */
+class ICDRegistrationDelegate
+{
+public:
+    static constexpr size_t KeySize = 16;
+    using ICDKey                    = FixedByteSpan<KeySize>;
+
+    /// The type of callback function that is called when symmetric key generation is completed.
+    typedef void (*OnSymmetricKeyGenerationCompleted)(void * context, CHIP_ERROR error, NodeId controllerNodeId, uint64_t subjectId,
+                                                      ICDKey key, Optional<ICDKey> verificationKey);
+
+    virtual ~ICDRegistrationDelegate() = default;
+
+    /**
+     * @brief
+     *   Called when the operational commissioning completed, the controller may register it with the ICD device.
+     * The delegate can generate the symmetric key and call commissioner->OnSymmetricKeyGenerated to complete ICD
+     * registration.
+     *   This delegate supports Commissioner Self-Registration and External Controller Registration flow.
+     *
+     * @param[in] device       The node id of the device.
+     * @param[in] onCompletion The callback object to be called when symmetric key generation is completed.
+     */
+    virtual void GenerateSymmetricKey(NodeId device, Callback::Callback<OnSymmetricKeyGenerationCompleted> * onCompletion) = 0;
+
+    /**
+     * @bried
+     *   Called when the ICD is ready for subscription. The callback is expected to trigger an async operation
+     * which subscription on the commissioner or external controllers.
+     *
+     * @param[in] device       The node id of the device.
+     * @param[in] icdCounter   The ICD Counter received from the device.
+     */
+    virtual void ReadyForSubscription(NodeId device, uint32_t icdCounter) = 0;
```

Comment:
The implementation is expected to (and should) persist the value. Especially for external controller registration flow.

Reply:
Agreed on the persistance but why would the persistance be done when are at the subscription establishment step?
Logically, the persistance of the ICDCounter value should be done when we receive the Registration reponse and create the associated entry.

Reply:
Renamed it to "RegistrationComplete" and will call it to `DeviceCommissioner::OnICDManagementRegisterClientResponse`.

The "ReadyForSubscription" should be triggerred by commissioning complete.

Reply:
When CommissioningComplete CB is called, I think we still need to provide signal to application saying this is ICD CommissioningComplete, application may construct its ICD subscription rather than the operations for non-ICD device..